### DISCUSSION TITLE
Which REST API is being used in Dropwizard 2.0.28, jakarta.ws.rs OR javax.ws.rs OR both?

### DISCUSSION THREAD

**Discussion by xjyang** [dropwizard/dropwizard#4794] [General]
I see there is a mixed usage of REST API in DW 2.0.28: in most (if not all) such as `dropwizard-auth` and `dropwizard-hibernate`, there are dependencies on `dropwizard-jersey`, which depends on `javax.ws.rs`. For example, the class `io.dropwizard.jersey.params.AbstractParam` has a lot of imports from` javax.ws`. In the meanwhile, DW also depends on` jakarta.ws.rs-api`. 

It looks like both `jakarta.ws.rs` and `javax.ws.rs` APIs are used in DW, right? Can I use just one of them? If yes, where/how to specify that?

I ask this question because I encountered a problem in my DW/Jersey based Web Application, which is caused by the mixed-use of the two:

```
java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    org/glassfish/jersey/servlet/WebComponent.<init>(Lorg/glassfish/jersey/servlet/WebConfig;Lorg/glassfish/jersey/server/ResourceConfig;)V @177: invokespecial
  Reason:
    Type 'org/glassfish/jersey/server/ResourceConfig' (current frame, stack[3]) is not assignable to 'javax/ws/rs/core/Application'
```

And I cannot find a way to fix this error. Any suggestion?

**Comment by zUniQueX** [comment]
After Java EE was donated to the Eclipse Foundation, the namespace `javax` had to be updated to `jakarta`. Therefore the package is now named `jakarta.ws.rs-api`. However, for backward compatibility, the classes of the version of this package used by Dropwizard use the `javax` namespace. The namespace of the classes will change in the next major update of the package.

Dropwizard will stay on the `javax` namespace in versions 2.x.x and move to the `jakarta` namespace in 3.0.0.

**Comment by zUniQueX** [comment]
I'm expecting a dependency conflict here. If your project pulls in a dependency of Jersey with a version `>=3.0.0`, the `ResourceConfig` class would extend `jakarta.ws.rs.core.Application`. Please check, that no transitive dependency contains an other Jersey version.

If that's not the case, then I would kindly ask for an example code snippet to reproduce this issue.

**Comment by joschi** [comment]
> It looks like both `jakarta.ws.rs` and `javax.ws.rs` APIs are used in DW, right?

As @zUniQueX already pointed out, Dropwizard 2.x is using Jersey 2.x which is an implementation for the API in the `javax.ws.rs` namespace:
```
# grep -r --no-filename 'import javax\.ws' . | wc -l
919

# grep -r --no-filename 'import jakarta\.ws' . | wc -l
0
```

The Maven artifact ID changed to Jakarta some time ago, so that there's understandably some confusion.

In your application, make sure to only include dependencies which implement the `javax.*` namespace.

**Comment by xjyang** [comment]
Thank you both for all the explanations! They are quite helpful.
